UHM-4640: Discharge should un-enroll from COVID-19 program

mogoodrich opened the pull request.

mogoodrich commented:
Thanks @lnball !

brandones commented on api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.Date;
+import java.util.List;
+
+public class CloseCovidProgramAction implements CustomFormSubmissionAction {
```
`ExitFromCovidProgramAction`? Hopefully the hospital keeps the program open :)

mogoodrich replied:
I can add "Patient", ie "ExitPatientFromCovidProgramAction"

brandones commented on api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+
+    private void closeCovidProgram(FormEntrySession formEntrySession, ProgramWorkflowService programWorkflowService, DispositionService dispositionService, ConceptService conceptService) {
```
```suggestion
    private void closeCovidProgram(FormEntrySession formEntrySession,
                                   ProgramWorkflowService programWorkflowService,
                                   DispositionService dispositionService,
                                   ConceptService conceptService) {
```

brandones commented on api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:
```diff
@@ -0,0 +1,112 @@
+        if (candidates != null) {
+
+            if (candidates.size() > 1) {
+                log.warn("More than one COVID program enrolment for patient " + patient.getId() + " on date " + encounterDate);
```
```suggestion
                log.warn("More than one COVID program enrollment for patient " + patient.getId() +
                         " on date " + encounterDate + ". Now unenrolling from all of them.");
```

brandones commented on api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:
```diff
@@ -0,0 +1,112 @@
+                // find the disposition
+                DispositionDescriptor dispositionDescriptor = dispositionService.getDispositionDescriptor();
+
+                Disposition disposition = null;
```
```suggestion
                Disposition disposition;
```

mogoodrich replied:
Without setting dispositon to null here you get a "variable might not have been initialized" warning on line 90 (which would happen if it never may any trips through the loop that starts on line 82)

brandones commented on api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:
```diff
@@ -0,0 +1,112 @@
+
+                // set program outcome based on dispositoion
+                if (disposition != null) {
+                    if (disposition.getUuid().equals(DISPOSITION_DEATH)) {
```
I wouldn't expect `disposition.getUuid()` to ever equal `"markPatientDead"`?

mogoodrich replied:
That's a good expectation, but it does... I don't know why that field is called uuid, it probably should be called key.  I should comment, however...

https://github.com/PIH/openmrs-config-zl/blob/master/configuration/pih/pih-dispositions-haiti.json#L50

brandones replied:
:open_mouth:

brandones commented:
@mogoodrich A few nits and one major issue

mogoodrich commented:
Thanks @brandones  I committed 2 of your suggested changes and commented about the other.

I renamed the class to something that makes a little more sense.

And I added a comment about the weird usage of uuid...
